src/support.js

```diff
@@ -328,158 +343,194 @@ function loadWebAssemblyModule(binary, loadAsync) {
-  // The static area consists of explicitly initialized data, followed by zero-initialized data.
-  // The latter may need zeroing out if the MAIN_MODULE has already used this memory area before
-  // dlopen'ing the SIDE_MODULE.  Since we don't know the size of the explicitly initialized data
-  // here, we just zero the whole thing, which is suboptimal, but should at least resolve bugs
-  // from uninitialized memory.
-  for (var i = memoryStart; i < memoryStart + memorySize; ++i) HEAP8[i] = 0;
-  // prepare env imports
-  var env = Module['asmLibraryArg'];
-  // TODO: use only __memory_base and __table_base, need to update asm.js backend
-  var table = Module['wasmTable'];
-  var oldTableSize = table.length;
-  env['__memory_base'] = env['gb'] = memoryStart;
-  env['__table_base'] = env['fb'] = oldTableSize;
-  var originalTable = table;
-  table.grow(tableSize);
-  assert(table === originalTable);
-  // zero-initialize memory and table
-  // TODO: in some cases we can tell it is already zero initialized
-  for (var i = env['__memory_base']; i < env['__memory_base'] + memorySize; i++) {
-    HEAP8[i] = 0;
-  }
-  for (var i = env['__table_base']; i < env['__table_base'] + tableSize; i++) {
-    table.set(i, null);
+
+  // shared libraries this module needs. We need to load them first, so that
+  // current module could resolve its imports. (see tools/shared.py
+  // WebAssembly.make_shared_library() for "dylink" section extension format)
+  var neededDynlibsCount = getLEB();
+  var neededDynlibs = [];
+  for (var i = 0; i < neededDynlibsCount; ++i) {
+    var nameLen = getLEB();
+    var nameUTF8 = binary.subarray(next, next + nameLen);
+    next += nameLen;
+    var name = UTF8ArrayToString(nameUTF8, 0);
+    neededDynlibs.push(name);
   }
-  // copy currently exported symbols so the new module can import them
-  for (var x in Module) {
-    if (!(x in env)) {
-      env[x] = Module[x];
+  if (neededDynlibs) {
+    if (flags.loadAsync) {
+      return Promise.all(neededDynlibs.map(function(dynNeeded) {
+        return loadDynamicLibrary(dynNeeded, flags);
+      })).then(function() {
+        return loadModule();
+      });
     }
+
+    neededDynlibs.forEach(function(dynNeeded) {
+      loadDynamicLibrary(dynNeeded, flags);
+    });
   }
-  // wasm dynamic libraries are pure wasm, so they cannot assist in
-  // their own loading. When side module A wants to import something
-  // provided by a side module B that is loaded later, we need to
-  // add a layer of indirection, but worse, we can't even tell what
-  // to add the indirection for, without inspecting what A's imports
-  // are. To do that here, we use a JS proxy (another option would
-  // be to inspect the binary directly).
-  var proxyHandler = {
-    'get': function(obj, prop) {
-      if (prop in obj) {
-        return obj[prop]; // already present
+
+  return loadModule();
```

Comment:
It's not bad to call this function before it's declared, but I think it's clearer when the call is afterward

Reply:
I thought about it initially and weighting all pros and cons decided that it was still clearier when the main logic was in the beginning - not scattered - and loadModule() implementation was following that.

Anyway I'm splitting this now upon your request with the following interdiff:

```diff
--- a/src/support.js
+++ b/src/support.js
@@ -356,21 +356,6 @@ function loadWebAssemblyModule(binary, flags) {
     var name = UTF8ArrayToString(nameUTF8, 0);
     neededDynlibs.push(name);
   }
-  if (neededDynlibs) {
-    if (flags.loadAsync) {
-      return Promise.all(neededDynlibs.map(function(dynNeeded) {
-        return loadDynamicLibrary(dynNeeded, flags);
-      })).then(function() {
-        return loadModule();
-      });
-    }
-
-    neededDynlibs.forEach(function(dynNeeded) {
-      loadDynamicLibrary(dynNeeded, flags);
-    });
-  }
-
-  return loadModule();
 
   // loadModule loads the wasm module after all its dependencies have been loaded.
   // can be called both sync/async.
@@ -532,6 +517,20 @@ function loadWebAssemblyModule(binary, flags) {
       return postInstantiation(instance);
     }
   }
+
+  // now load needed libraries and the module itself.
+  if (flags.loadAsync) {
+    return Promise.all(neededDynlibs.map(function(dynNeeded) {
+      return loadDynamicLibrary(dynNeeded, flags);
+    })).then(function() {
+      return loadModule();
+    });
+  }
+
+  neededDynlibs.forEach(function(dynNeeded) {
+    loadDynamicLibrary(dynNeeded, flags);
+  });
+  return loadModule();
 }
 Module['loadWebAssemblyModule'] = loadWebAssemblyModule;
```